src/main/java/cn/nukkit/Player.java

```diff
@@ -820,27 +815,26 @@ protected void sendNextChunk() {
                 }
             }
         }
-        if (this.chunkLoadCount >= this.spawnThreshold && !this.spawned && this.teleportPosition == null) {
-            this.doFirstSpawn();
+        if (this.spawnChunkLoadCount != -1 && ++this.spawnChunkLoadCount >= this.spawnThreshold) {
+            this.sendPlayStatus(PlayStatusPacket.PLAYER_SPAWN);
+            this.spawnChunkLoadCount = -1;
         }
         Timings.playerChunkSendTimer.stopTiming();
     }
 
     protected void doFirstSpawn() {
+        if (this.spawned) {
+            return;
+        }
         this.spawned = true;
+        this.setImmobile(false);
 
-        this.setEnableClientCommand(true);
-
-        this.getAdventureSettings().update();
-
-        this.sendPotionEffects(this);
-        this.sendData(this);
-        this.inventory.sendContents(this);
-        this.inventory.sendArmorContents(this);
-
-        SetTimePacket setTimePacket = new SetTimePacket();
-        setTimePacket.time = this.level.getTime();
-        this.dataPacket(setTimePacket);
+        if (this.hasPermission(Server.BROADCAST_CHANNEL_USERS)) {
```

Comment:
Is there any reason for permission subscription to be moved here?

Reply:
This was causing an occasional client-sided crash due to `TextPackets` being sent to players at bad times during the login sequence.